File: skyeye-statistics/src/main/scala/com/jthink/skyeye/statistics/launcher/Launcher.scala
package com.jthink.skyeye.statistics.launcher

import com.jthink.skyeye.statistics.properties.TaskProperties
import com.jthink.skyeye.statistics.task.TestTask
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.context.annotation.{ComponentScan, PropertySource}

/**
  * JThink@JThink
  *
  * 项目启动器
  *
  * @author JThink
  * @version 0.0.1
  */
@SpringBootApplication
@ComponentScan(basePackages = Array("com.jthink.teemo"))
//@PropertySource(Array("file:/the/file/location/teemo/teemo.properties"))
@PropertySource(Array("classpath:properties/teemo.properties"))
@EnableConfigurationProperties(Array(classOf[TaskProperties]))
class Launcher {
}

object Launcher {
  private val TASK_MAPPING = Map("test" -> classOf[TestTask])
  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[Launcher])

  def main(args: Array[String]): Unit = {
    if (args.length != 1) {
      LOGGER.info("参数不对, 参数如下:")
      LOGGER.info("参数1: 需要执行的task name")
    } else {
      val t = args(0).toString
      if (!TASK_MAPPING.contains(t)) {
        LOGGER.info("task 不存在")
      } else {
        val context: ConfigurableApplicationContext = new SpringApplicationBuilder(classOf[Launcher]).run(args: _*)
        LOGGER.info("teemo start successfully")
        val task = context.getBean(TASK_MAPPING.get(t).get)
        // 开始执行具体的业务
        task.doTask()
      }
    }
  }
}

File: skyeye-collector/skyeye-collector-metrics/src/main/java/com/jthink/skyeye/collector/metrics/launcher/Launcher.java
package com.jthink.skyeye.collector.metrics.launcher;

import com.jthink.skyeye.collector.core.hook.ShutdownHookRunner;
import com.jthink.skyeye.collector.core.task.Task;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.boot.logging.LoggingApplicationListener;
import org.springframework.context.ApplicationListener;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.PropertySource;

import java.util.Iterator;
import java.util.Set;

/**
 * JThink@JThink
 *
 * @author JThink
 * @version 0.0.1
 * @desc 项目启动器
 * @date 2016-08-24 18:31:48
 */
@SpringBootApplication
@EnableAutoConfiguration
@ComponentScan(basePackages={"com.jthink.skyeye.collector.core", "com.jthink.skyeye.collector.metrics", "com.jthink.skyeye.data.rabbitmq", "com.jthink.skyeye.data.jpa"})
@PropertySource("file:/opt/jthink/jthink-config/skyeye/collector/collector-metrics.properties")
//@PropertySource("classpath:properties/collector-metrics.properties")
public class Launcher {

    private static final Logger LOGGER = LoggerFactory.getLogger(Launcher.class);

    public static void main(String[] args) {
        SpringApplicationBuilder builder = new SpringApplicationBuilder(Launcher.class);
        Set<ApplicationListener<?>> listeners = builder.application().getListeners();
        for (Iterator<ApplicationListener<?>> it = listeners.iterator(); it.hasNext();) {
            ApplicationListener<?> listener = it.next();
            if (listener instanceof LoggingApplicationListener) {
                it.remove();
            }
        }
        builder.application().setListeners(listeners);
        ConfigurableApplicationContext context = builder.run(args);
        LOGGER.info("collector metrics start successfully");

        KafkaConsumer kafkaConsumer = (KafkaConsumer<byte[], String>) context.getBean("kafkaConsumer");
        Task task = (Task) context.getBean("metricsTask");

        // 优雅停止项目
        Runtime.getRuntime().addShutdownHook(new ShutdownHookRunner(kafkaConsumer, task));
        task.doTask();
    }

}
